google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py

```diff
@@ -191,9 +191,20 @@ def ack_deadline(self):
         Returns:
             int: The ack deadline.
         """
-        target = min([self._last_histogram_size * 2, self._last_histogram_size + 100])
-        if len(self.ack_histogram) > target:
-            self._ack_deadline = self.ack_histogram.percentile(percent=99)
+        target_size = min(
+            self._last_histogram_size * 2, self._last_histogram_size + 100
+        )
+        hist_size = len(self.ack_histogram)
+
+        if hist_size > target_size:
+            self._last_histogram_size = hist_size
+            p99 = self.ack_histogram.percentile(percent=99)
+
+            if self.flow_control.max_extension_period > 0:
+                self._ack_deadline = min(p99, self.flow_control.max_extension_period)
+            else:
+                self._ack_deadline = p99
```

Comment:
As I already mentioned in original issue, I see an error here: `_last_histogram_size` is never changed. According to the method docs, it should be implemented like this. We're remembering histogram size only if we got enough data. Otherwise we can get in a situation when data added with very small portions a lot of times - in total we got a lot of data, but we're not processing it, 'cause growth steps were small

Reply:
Yes, that was a bug. Good catch!